### PR TITLE
Simplify and fix decision engine's handling of skolem definitions

### PR DESCRIPTION
**Pull request by ajreynol**
This PR changes the front end of prop engine to distinguish input formulas from skolem definitions, which is required by the decision engine.  This PR breaks the dependency of PropEngine on the AssertionsPipeline, as now the determining of whether an input formula is a skolem definition is handled externally, in SmtSolver.

With this PR, we should not be required to satisfy skolem definitions that are not relevant based on the techniques already implemented in the decision engine.  Currently, we are not distinguishing input formulas from skolem definitions properly, meaning we assert more literals than we need to.

This also simplifies related interfaces within decision engine.

We should test this PR with --decision=justification on SMT-LIB.

### CODE REVIEW

**Comment by HanielB** [comment]
Results for a full SMT-LIB run: 
```
   DIRECTORY |                                                   master |                                                   branch |
   Benchmark |  Stat   Slvd    Tot    To Mo    Err    Cpu[s]    Mem[MB] |  Stat   Slvd    Tot    To Mo    Err    Cpu[s]    Mem[MB] |
         ABV |    ok     17    169     0  0      0       1.5      218.4 |    ok     17    169     0  0      0       1.5      215.4 |
       ABVFP |    ee      0     30     0  0     30       0.0        0.0 |    ee      0     30     0  0     30       0.0        0.0 |
    ABVFPLRA |    ee      0     75     0  0     75       0.0        0.0 |    ee      0     75     0  0     75       0.0       44.9 |
        ALIA |    ok     35     42     0  0      0       1.8      171.5 |    ok     35     42     0  0      0       1.7      173.4 |
  AUFBVDTLIA |    ok    411   1708   864  0      0   51995.5   352857.1 |    ok    411   1708   864  0      0   51993.1   352365.1 |
  AUFBVDTNIA |    ok      0      1     0  0      0       0.0        0.0 |    ok      0      1     0  0      0       0.0        0.0 |
    AUFDTLIA |    ok    635    728     0  0      0       0.1       86.3 |    ok    635    728     0  0      0       0.0        0.0 |
   AUFDTLIRA |    ee  10518  12196   121  0    519    7959.3    73423.1 |    ee  10518  12196   121  0    519    7943.6    62671.1 |
   AUFDTNIRA |    ok    327    471    73  0      0    4444.5    86564.8 |    ok    327    471    73  0      0    4445.6    88263.6 |
 AUFFPDTLIRA |    ee      0    154     0  0    154       0.0        0.0 |    ee      0    154     0  0    154       0.0        0.0 |
 AUFFPDTNIRA |    ee      0     13     0  0     13       0.0        0.0 |    ee      0     13     0  0     13       0.0        0.0 |
      AUFLIA |    ok   2026   3276   440  0      0   27491.3   179721.8 |    ok   2027   3276   439  0      0   27443.0   177816.3 |
     AUFLIRA |    ok  19732  20011   125  0      0   10398.0    48834.6 |    ok  19732  20011   125  0      0   10491.6    41419.7 |
      AUFNIA |    ok      0      3     0  0      0       0.0        0.0 |    ok      0      3     0  0      0       0.0        0.0 |
     AUFNIRA |    ok   1033   1480   400  0      0   24342.3    86434.7 |    ok   1033   1480   400  0      0   24338.1    86002.1 |
          BV |    ok   5457   5846   389  0      0   25016.1   246843.8 |    ok   5457   5846   389  0      0   25135.3   240623.4 |
        BVFP |    ee      0    224     0  0    224       0.0        0.0 |    ee      0    224     0  0    224       0.0        0.0 |
     BVFPLRA |    ee      0    185     0  0    185       0.0        0.0 |    ee      0    185     0  0    185       0.0        0.0 |
          FP |    ee      0   2484     0  0   2484       0.0        0.0 |    ee      0   2484     0  0   2484       0.0     1653.3 |
       FPLRA |    ee      0     27     0  0     27       0.0        0.0 |    ee      0     27     0  0     27       0.0        0.0 |
         LIA |    ok    577    607    30  0      0    1861.8     6587.1 |    ok    577    607    30  0      0    1861.5     6463.0 |
         LRA |    ok   1876   2419   543  0      0   33434.2    45294.6 |    ok   1876   2419   543  0      0   33440.3    44829.3 |
         NIA |    ok     17     20     3  0      0     182.3      248.8 |    ok     17     20     3  0      0     182.8      247.2 |
         NRA |    ok   3782   3813    26  0      0    1587.6     5527.2 |    ok   3782   3813    26  0      0    1586.3     4846.4 |
      QF_ABV |    ok  14395  15084   689  0      0   72630.0   475516.2 |    ok  14396  15084   688  0      0   72685.6   478407.5 |
    QF_ABVFP |    ee      0  18093     0  0  18093       0.0        0.0 |    ee      0  18093     0  0  18093       0.0      790.5 |
 QF_ABVFPLRA |    ee      0     74     0  0     74       0.0        0.0 |    ee      0     74     0  0     74       0.0        0.0 |
     QF_ALIA |    ok    120    126     6  0      0     817.2     2688.6 |    ok    120    126     6  0      0     831.2     2518.1 |
     QF_ANIA |    ok     86     95     9  0      0     842.1     3069.8 |    ok     86     95     9  0      0     849.6     2966.5 |
    QF_AUFBV |    ok     32     56    24  0      0    1457.1    36945.4 |    ok     32     56    24  0      0    1455.8    36054.4 |
   QF_AUFLIA |    ok   1298   1303     5  0      0     867.9     9198.3 |    ok   1299   1303     4  0      0     861.5     8673.5 |
   QF_AUFNIA |    ok     17     17     0  0      0     125.6      831.3 |    ok     17     17     0  0      0     117.2      834.9 |
       QF_AX |    ok    551    551     0  0      0     335.1     4101.2 |    ok    551    551     0  0      0     335.6     3998.4 |
       QF_BV |    ok  36941  41713  4764  8      0  367855.2  4657576.7 |    ok  36953  41713  4752  8      0  367539.8  4653962.1 |
     QF_BVFP |    ee      0  17231     0  0  17231       0.0        0.0 |    ee      0  17231     0  0  17231       0.0       10.9 |
  QF_BVFPLRA |    ee      0    168     0  0    168       0.0        0.0 |    ee      0    168     0  0    168       0.0        0.0 |
       QF_DT |    ok   8000   8000     0  0      0      33.2     2192.1 |    ok   8000   8000     0  0      0      35.7      305.0 |
       QF_FP |    ee      0  40318     0  0  40318       0.0      759.1 |    ee      0  40318     0  0  40318       0.0      186.3 |
    QF_FPLRA |    ee      0     57     0  0     57       0.0        0.0 |    ee      0     57     0  0     57       0.0        0.0 |
      QF_IDL |    ok   1127   2193  1066  0      0   71808.2   400032.4 |    ok   1130   2193  1063  0      0   71856.2   399735.8 |
      QF_LIA |    ok   3655   6947  3292  0      0  224694.5  2089995.2 |    ok   3651   6947  3296  0      0  224999.5  2092336.6 |
     QF_LIRA |    ok      3      7     4  0      0     241.6     1040.4 |    ok      3      7     4  0      0     241.7     1039.0 |
      QF_LRA |    ok   1195   1648   453  0      0   32084.7   147022.2 |    ok   1198   1648   450  0      0   32056.9   146659.1 |
      QF_NIA |    ok   7773  23876  6582  0      0  483066.1  1620610.3 |    ok   7768  23876  6584  0      0  483792.1  1613825.9 |
     QF_NIRA |    ok      1      3     2  0      0     120.1      549.3 |    ok      1      3     2  0      0     120.2      584.9 |
      QF_NRA |    ok   8503  11489  2801  0      0  179510.0   550816.0 |    ok   8502  11489  2801  0      0  179508.5   546006.0 |
      QF_RDL |    ok    138    255   117  0      0    7875.5    23878.1 |    ok    138    255   117  0      0    7863.0    23912.1 |
        QF_S |    ee   1028   2319     0  0   1291       6.2      559.1 |    ee   1028   2319     0  0   1291       6.0      569.2 |
     QF_SLIA |    ee  48073  67584   344  0  19167   22173.8   113345.9 |    ee  48074  67584   344  0  19166   22194.4   114953.6 |
       QF_UF |    ok   7352   7457   105  0      0   13511.9   124411.5 |    ok   7353   7457   104  0      0   13499.1   116278.4 |
     QF_UFBV |    ok   1169   1234    65  0      0    7646.4    98615.8 |    ok   1172   1234    62  0      0    7250.8    95024.3 |
     QF_UFFP |    ee      0      2     0  0      2       0.0        0.0 |    ee      0      2     0  0      2       0.0        0.0 |
    QF_UFIDL |    ok    278    428   150  0      0    9815.5    19738.7 |    ok    278    428   150  0      0    9782.6    19754.0 |
    QF_UFLIA |    ok    509    583    74  0      0    5044.4     8777.8 |    ok    509    583    74  0      0    5061.8     8891.1 |
    QF_UFLRA |    ok   1270   1284    14  0      0    1416.5    57735.6 |    ok   1270   1284    14  0      0    1455.0    54926.9 |
    QF_UFNIA |    ok    445    478     2  0      0     167.8     2178.6 |    ok    445    478     2  0      0     172.7     1913.7 |
    QF_UFNRA |    ok     13     37    24  0      0    1443.0     3804.9 |    ok     13     37    24  0      0    1443.3     3819.6 |
          UF |    ok   3217   7668  3714  0      0  225363.0   698529.0 |    ok   3218   7668  3713  0      0  225387.4   695520.8 |
        UFBV |    ok     55    121    11  0      0    1038.3    20919.2 |    ok     55    121    11  0      0    1036.0    21201.2 |
        UFDT |    ok   2108   4527  2040  0      0  123693.4   436488.2 |    ok   2109   4527  2039  0      0  123668.9   434419.2 |
     UFDTLIA |    ok     73    327   243  0      0   14657.6    57083.5 |    ok     73    327   243  0      0   14657.8    56859.9 |
    UFDTLIRA |    ee   8327  10743    19  0    595    1284.8    33004.8 |    ee   8327  10743    19  0    595    1282.2    24495.0 |
     UFDTNIA |    ok      1      1     0  0      0       0.2       31.9 |    ok      1      1     0  0      0       0.2       32.0 |
    UFDTNIRA |    ok   1625   1839    86  0      0    5395.1   176372.6 |    ok   1625   1839    86  0      0    5394.9   182528.6 |
  UFFPDTLIRA |    ee      0    933     0  0    933       1.5      170.2 |    ee      0    933     0  0    933       0.0       17.3 |
  UFFPDTNIRA |    ee      0     21     0  0     21       0.2       10.0 |    ee      0     21     0  0     21       0.0       13.6 |
       UFIDL |    ok     56     68     5  0      0     301.4     1397.5 |    ok     56     68     5  0      0     301.5     1421.9 |
       UFLIA |    ok   7381  10137  2043  0      0  127890.8   570481.1 |    ok   7381  10137  2042  0      0  127882.2   564427.7 |
       UFLRA |    ok     10     15     0  0      0       0.1       35.0 |    ok     10     15     0  0      0       0.1       24.6 |
       UFNIA |    ok   5497  13509  2948  0      0  189975.8  2197840.3 |    ok   5504  13509  2939  0      0  189786.2  2176637.4 |
    __totals |    ee 218765 376601 34715  8 101661 2383908.2 15781167.6 |    ee 218790 376601 34684  8 101660 2384277.5 15694171.7 |```

File: src/prop/theory_proxy.cpp [hunk 1]
```diff
@@ -156,16 +156,14 @@ void TheoryProxy::spendResource(ResourceManager::Resource r)
   d_theoryEngine->spendResource(r);
 }
 
-bool TheoryProxy::isDecisionRelevant(SatVariable var) {
-  return d_decisionEngine->isRelevant(var);
-}
+bool TheoryProxy::isDecisionRelevant(SatVariable var) { return true; }
```
Comment (HanielB) [inline src/prop/theory_proxy.cpp:159]: This and the change below look a bit odd. :) Could this affect `decision==internal`? Moreover, could you comment on these changes and update the doc (currently non-existent hehe) in the headers? OTOH, if these methods are to behave like this then we should change the SAT solver to not rely on them as well.

File: src/decision/decision_engine.h [hunk 2]
```diff
@@ -142,23 +127,18 @@ class DecisionEngine {
   // External World helping us help the Strategies
 
   /**
-   * Add a list of assertions, as well as lemmas coming from preprocessing
-   * (ppLemmas) and pairwise the skolems they constrain (ppSkolems).
+   * Add that assertion is an (input) assertion, not corresponding to a
```
Comment (HanielB) [inline src/decision/decision_engine.h:130]: This phrasing seems a bit odd. You mean in the sense of notifying, maybe change accordingly?

File: src/prop/theory_proxy.cpp (hunk 1 above)
Comment (ajreynol) [reply src/prop/theory_proxy.cpp:159]: This change preserves behavior, isRelevant always returned true.  We could do the further refactoring to not make SAT solver rely on this method. I considered doing this on this PR but decided to wait for now.

Comment (HanielB) [reply src/prop/theory_proxy.cpp:159]: OK, sounds good.

File: src/decision/decision_engine.h (hunk 2 above)
Comment (HanielB) [reply src/decision/decision_engine.h]: BTW this pattern happens is several other places in the PR, could you change those as well?

**Review by HanielB — APPROVED** [review]

**Pull request merged**